### PR TITLE
policy: Per route authorization

### LINKED ISSUE #8890: policy: Allow AuthorizationPolicies to target HTTPRoutes
Now that the policy controller can read HTTPRoute resources and serve them in proxy configurations, we should update the policy controller to allow AuthorizationPolicies to target HTTPRoute instances (in the same namespace) to augment Server authorizations.

We should add API, admission, and e2e tests to confirm this behavior.

### PR DESCRIPTION
**Pull request by adleong**
Fixes #8890

When building the `InboundHttpRoute`, we find all authorizations which target that route and copy the associated authentications onto the route and return them in the api.  This allows AuthorizationPolicies to target HttpRoutes.

We also add admission, api, and e2e tests.

### CODE REVIEW

File: policy-controller/k8s/index/src/authorization_policy.rs [hunk 1]
```diff
@@ -65,6 +66,9 @@ fn target(t: LocalTargetRef) -> Result<Target> {
     if t.targets_kind::<k8s::Namespace>() {
         return Ok(Target::Namespace);
     }
+    if t.targets_kind::<k8s_gateway_api::HttpRoute>() {
```
Comment (hawkw) [inline policy-controller/k8s/index/src/authorization_policy.rs:69]: very tiny style nit (sorry): normally, i think we would put a line of whitespace between these
```suggestion
    }
    
    if t.targets_kind::<k8s_gateway_api::HttpRoute>() {
```

File: policy-controller/k8s/index/src/index.rs [hunk 2]
```diff
@@ -1172,11 +1176,71 @@ impl PolicyIndex {
...
+                        authorizationpolicy = %name,
+                        route = %route_name,
+                        target = %name,
+                        "AuthorizationPolicy does not target HttpRoute",
+                    );
+                    continue;
+                }
+            } else {
+                continue;
+            }
```
Comment (hawkw) [inline policy-controller/k8s/index/src/index.rs:1201]: the control flow here feels a _little_ hard to follow, with the multiple `continue`s...i'm not really sure if there is a way to make it much simpler, so i'm fine with it as-is, but i don't _love_ it...

File: policy-controller/k8s/index/src/index.rs [hunk 3]
```diff
@@ -1172,11 +1176,71 @@ impl PolicyIndex {
...
 
-    fn http_routes(&self, server_name: &str) -> HashMap<String, InboundHttpRoute> {
+    fn route_client_authzs(
+        &self,
+        route_name: &str,
+        authentications: &AuthenticationNsIndex,
+    ) -> HashMap<AuthorizationRef, ClientAuthorization> {
+        let mut authzs = HashMap::default();
+
+        for (name, spec) in self.authorization_policies.iter() {
```
Comment (hawkw) [inline policy-controller/k8s/index/src/index.rs:1186]: style nit, not important: `&HashMap` implements `IntoIterator`, so i believe this could just be 
```suggestion
        for (name, spec) in self.authorization_policies {
```

File: policy-controller/k8s/index/src/index.rs [hunk 4]
```diff
@@ -1172,11 +1176,71 @@ impl PolicyIndex {
...
+                continue;
+            }
+
+            tracing::trace!(
+                ns = %self.namespace,
+                authorizationpolicy = %name,
+                route = %route_name,
+                "AuthorizationPolicy targets HttpRoute",
+            );
+            tracing::trace!(authns = ?spec.authentications);
```
Comment (hawkw) [inline policy-controller/k8s/index/src/index.rs:1209]: nit, TIOLI: should this just be a single event?
```suggestion
            tracing::trace!(
                ns = %self.namespace,
                authorizationpolicy = %name,
                route = %route_name,
                authns = ?spec.authentications,
                "AuthorizationPolicy targets HttpRoute",
            );
```

File: policy-controller/k8s/index/src/index.rs (hunk 2 above)
Comment (hawkw) [reply policy-controller/k8s/index/src/index.rs:1201]: (side note, in Rust 1.64, we'll be able to use the `if let` chains that stabilized in https://github.com/rust-lang/rust/pull/94927#event-7007028976, which might simplify this a little)

File: policy-controller/k8s/index/src/authorization_policy.rs (hunk 1 above)
Comment (olix0r) [reply policy-controller/k8s/index/src/authorization_policy.rs:69]: this wouldn't be consistent with the rest of this function. so if we make ws changes here we should do so above.

File: policy-controller/k8s/index/src/index.rs (hunk 2 above)
Comment (olix0r) [reply policy-controller/k8s/index/src/index.rs:1201]: probably clearer with a match:

```suggestion
            match &spec.target {
                authorization_policy::Target::HttpRoute(n) if n == route_name => {}
                _ => {
                    tracing::trace!(
                        ns = %self.namespace,
                        authorizationpolicy = %name,
                        route = %route_name,
                        target = %name,
                        "AuthorizationPolicy does not target HttpRoute",
                    );
                    continue;
                }
            }
```

File: policy-controller/k8s/index/src/index.rs (hunk 4 above)
Comment (olix0r) [reply policy-controller/k8s/index/src/index.rs:1209]: I'm not sure that this is an improvement. `authns` might contain several/many largish datastructures that will make the log line less readable.

File: policy-test/tests/api.rs [hunk 5]
```diff
@@ -334,7 +334,7 @@ async fn server_with_http_route() {
 
         // Create an http route that refers to the `linkerd-admin` server (by
         // name) and ensure that the update now reflects this route.
-        create(
+        let route = create(
```
Comment (olix0r) [inline policy-test/tests/api.rs:337]: i'm probably missing something but is this `route` variable actually used?

File: policy-controller/k8s/index/src/index.rs [hunk 6]
```diff
@@ -1142,6 +1142,10 @@ impl PolicyIndex {
                     }
                 }
                 authorization_policy::Target::Namespace => {}
+                authorization_policy::Target::HttpRoute(_) => {
+                    // Skip the policy if it targets an HttpRoute.
+                    continue;
```
Comment (olix0r) [inline policy-controller/k8s/index/src/index.rs:1147]: This comment is a little redundant. It might be helpful, though, _why_ it's okay to skip the policy. I.e., explain where these policies are going to be handled.

**Review by olix0r — COMMENTED** [review]
This is looking pretty good to me! I think we need some deeper testing--I'll try to do some manual testing in the meantime.

File: policy-test/tests/e2e_authorization_policy.rs [hunk 7]
```diff
@@ -48,6 +50,63 @@ async fn meshtls() {
...
+        assert_eq!(
+            allowed_status, 0,
+            "curling allowed route must contact nginx"
+        );
+        assert_ne!(
+            denied_status, 0,
+            "curl which does not match route must not contact nginx"
+        );
+    })
+    .await;
```
Comment (olix0r) [inline policy-test/tests/e2e_authorization_policy.rs:107]: There are probably a few more states worth considering here:

* allowed by a server-scoped authorization
* allowed by a server-scoped authorization but not present in routes (should get a 404).

not a blocker

File: policy-test/tests/e2e_authorization_policy.rs [hunk 8]
```diff
@@ -48,6 +50,63 @@ async fn meshtls() {
...
+        );
+
+        let curl = curl::Runner::init(&client, &ns).await;
+        let (allowed, denied) = tokio::join!(
+            curl.run(
+                "curl-allowed",
+                "http://nginx/allowed",
+                LinkerdInject::Enabled
+            ),
+            curl.run("curl-denied", "http://nginx/denied", LinkerdInject::Enabled),
```
Comment (olix0r) [inline policy-test/tests/e2e_authorization_policy.rs:95]: This should probably target an allowed resource but not have sufficient authorization to show that authorization actually works:

```suggestion
            curl.run("curl-denied", "http://nginx/allowed", LinkerdInject::Disabled),
```

File: policy-test/src/lib.rs [hunk 9]
```diff
@@ -167,13 +167,13 @@ where
...
-    if let Err(err) = res {
-        std::panic::resume_unwind(err.into_panic());
-    }
+    // tracing::debug!(ns = %ns.name_unchecked(), "Deleting");
+    // api.delete(&ns.name_unchecked(), &kube::api::DeleteParams::background())
+    //     .await
+    //     .expect("failed to delete Namespace");
+    // if let Err(err) = res {
+    //     std::panic::resume_unwind(err.into_panic());
+    // }
```
Comment (olix0r) [inline policy-test/src/lib.rs:176]: Can we re-enable this?

File: policy-test/tests/e2e_authorization_policy.rs [hunk 10]
```diff
@@ -48,6 +50,73 @@ async fn meshtls() {
...
+        );
+
+        let curl = curl::Runner::init(&client, &ns).await;
+        let (allowed, denied, unauth) = tokio::join!(
+            curl.run(
+                "curl-allowed",
+                "http://nginx/allowed",
+                LinkerdInject::Enabled
+            ),
+            curl.run("curl-denied", "http://nginx/denied", LinkerdInject::Enabled),
```
Comment (olix0r) [inline policy-test/tests/e2e_authorization_policy.rs:95]: Sorry, I still think this is a little off: this case is hitting an endpoint on no route. we probably want an additional route object that matches `/denied` (or an empty route to match all endpoints?). OR we should rename "denied" to something like "no-route"

**Review by hawkw — APPROVED** [review]
